### ISSUE TITLE
Unnecessary use of floating point operations when using fuse_reply_entry or fuse_add_direntry_plus

### ISSUE THREAD

**Issue by davidgaleano** [libfuse/libfuse#727]
The structure `fuse_entry_param` requires storing the timeout values as `double`s which internally are converted to integers when filling the `fuse_entry_out` structure that is sent back to the kernel.
This conversion from floating point values back to integer values is unfortunate when the application using libfuse can already provide the timeouts as integer seconds and nanoseconds, which is our case, we go from integers to floating point and back to integers.
It would be great if there was an alternative way of passing those timeouts using integer values, perhaps even to be able to fill the internal structure `fuse_entry_out` directly instead of going to the transient representation of a `stat` structure, a `fuse_reply_entry_raw` perhaps?

**Comment by Nikratio** [comment]
I didn't check the code, but if you're right then pull requests are welcome.

**Comment by bsbernd** [comment]
Entirely agreed, we are setting 200ns and that gets rather close to resolution of double. The other part that I don't like at all is that struct fuse_entry_param has 'fuse_ino_t ino', while fuse_entry_out and kernel actually treat this as nodeid (inode identifier) - totally confusing to call this 'ino', as one could assume that is the same as st_ino in struct stat.

**Comment by bsbernd** [comment]
@davidgaleano could you review https://github.com/libfuse/libfuse/pull/1032 ?

**Comment by davidgaleano** [comment]
That change is an improvement but, to be honest, it is rather messy and I think it actually complicates the library unnecessarily. Having to pass `timeout_as_double` around to know how to interpret the incoming entry parameter is unfortunate, at a minimum I would suggest to get rid of it and to assume that **all** internal APIs will use `timespec`, the external API can still use doubles or timespec according to the `#define` but when using doubles they will always be converted to call the `timespec` internal functions.

Also, I am concerned about existing programs dynamically linking to libfuse, this new version will break the ABI, right? `fuse_entry_param` is bigger because of the union with the timespec on it, and I am not sure if the new static inline functions will appear as part of the external symbols. I guess you were planning to increase the library version?

In any case, my hope was for a new set of functions and structures that totally avoided the need for `fill_entry` or `convert_stat`, basically let the application fill the data already in the format that the kernel requires, without the need for intermediate copies. That new set of functions would be leaner and simpler than the current one and it will not be hampered by having to support older versions of the API.

**Comment by davidgaleano** [comment]
The point is to not need timeout_as_double, internal functions should all work with timespec and the external functions for the double version are just thin wrappers that do the conversion to timespec so they can call the internal ones.

Ah, sorry, did not notice compact.c, I presumed incorrectly it was used only for testing.

**Comment by bsbernd** [comment]
Yes, I see your point about  timeout_as_double, issue is still that conversion from struct timespec to double cannot be done with a simple cast. So either we need to define this struct two times and then copy between the structs as in compat.c - even more overhead for the older double users or need to memcpy it - either one is overhead.

**Comment by davidgaleano** [comment]
Users that do not rebuild their applications after the new library is installed are going to pay the conversion cost on compat.c anyway, so we are not avoiding any overhead for them with the current proposal as it is. Their programs would be a tiny bit slower for no added benefit.
Users that do rebuild their applications will still pay the added cost of passing timeout_as_double around (and its runtime check to select the correct branch when translating the entry).
In any case we are adding overhead no matter what version is used. You could decide which combination is the one you want to favor and reduce its overhead.

I still think the best solution is a new API that exposes the kernel structures and does not require fill_entry and convert_stat. To avoid duplication you can make the old API call the new one after it uses fill_entry and convert_stat. The old API would be just a shim layer on top of the new one. I do understand this may be a big change that requires a lot of restructuring.

In any case, I really appreciate your effort and I would take advantage of anything you provide.

**Comment by bsbernd** [comment]
I'm closing it - no need to have wish-list issues open, if nobody is going to work on it. The approach I took is not welcomed (I closed PR #1032 therefore) and I certainly don't have time to work on another API.
